lib/spack/spack/spec.py

```diff
@@ -1196,7 +1196,9 @@ def virtual(self):
     @staticmethod
     def is_virtual(name):
         """Test if a name is virtual without requiring a Spec."""
-        return (name is not None) and (not spack.repo.path.exists(name))
+        return (name is not None and
+                (not spack.repo.path.exists(name) or
+                 spack.repo.path.get_pkg_class(name).virtual))
```

Comment:
It's not clear who is ultimately responsible for determining if a package is virtual.  I think this logic needs to get folded into the repo -- right now this consults the repo *and* the class, but there should really be one source of truth.  Can you fold this logic into `is_virtual`?  I realize that does the caching, so it may take some thought to figure out how to structure this.

Reply:
It definitely did take some work, but it proved doable. I had to add a parameter for whether to allow use of the index for a particular call to `repo.path.is_virtual`.